# Reference

Introduction:

[1] B. Tan, H. Liu, S. Zhang, S. R. da Silva, L. Zhang, J. Meng et al., "Viral and cellular N(6)-methyladenosine and N(6),2'-O-dimethyladenosine epitranscriptomes in the KSHV life cycle," (in eng), Nat Microbiol, vol. 3, no. 1, pp. 108-120, 2018, doi: 10.1038/s41564-017-0056-8. [[paper](https://pubmed.ncbi.nlm.nih.gov/29109479/)]

[2] M. E. Hess, S. Hess, K. D. Meyer, L. A. Verhagen, L. Koch, H. S. Brönneke et al., "The fat mass and obesity associated gene (Fto) regulates activity of the dopaminergic midbrain circuitry," (in eng), Nat Neurosci, vol. 16, no. 8, pp. 1042-8, Aug 2013, doi: 10.1038/nn.3449. [[paper](https://www.nature.com/articles/nn.3449)]

M0:

[3] Y. Tang, K. Chen, B. Song, J. Ma, X. Wu, Q. Xu et al., "m6A-Atlas: a comprehensive knowledgebase for unraveling the N6-methyladenosine (m6A) epitranscriptome," Nucleic Acids Res, vol. 49, no. D1, pp. D134-D143, 2020, doi: 10.1093/nar/gkaa692. [[paper](https://academic.oup.com/nar/article/49/D1/D134/5895329)]

[4] K. Chen, B. Song, Y. Tang, Z. Wei, Q. Xu, J. Su et al., "RMDisease: a database of genetic variants that affect RNA modifications, with implications for epitranscriptome pathogenesis," Nucleic Acids Res, vol. 49, no. D1, pp. D1396-D1404, 2020, doi: 10.1093/nar/gkaa790. [[paper](https://academic.oup.com/nar/article/49/D1/D1396/5917656)]

[5] K. Chen, Z. Wei, Q. Zhang, X. Wu, R. Rong, Z. Lu et al., "WHISTLE: a high-accuracy map of the human N6-methyladenosine (m6A) epitranscriptome predicted using a machine learning approach," Nucleic Acids Res, vol. 47, no. 7, pp. e41-e41, 2019, doi: 10.1093/nar/gkz074. [[paper](https://academic.oup.com/nar/article/47/7/e41/5319125?login=true)] 

M1:

[6] D. Kim, J. M. Paggi, C. Park, C. Bennett, and S. L. Salzberg, "Graph-based genome alignment and genotyping with HISAT2 and HISAT-genotype," Nature Biotechnology, vol. 37, no. 8, pp. 907-915, 2019/08/01 2019, doi: 10.1038/s41587-019-0201-4. [[paper](https://pubmed.ncbi.nlm.nih.gov/31375807/)]

[7] H. Li, B. Handsaker, A. Wysoker, T. Fennell, J. Ruan, N. Homer et al., "The Sequence Alignment/Map format and SAMtools," (in eng), Bioinformatics, vol. 25, no. 16, pp. 2078-9, Aug 15 2009, doi: 10.1093/bioinformatics/btp352.[[paper](https://pubmed.ncbi.nlm.nih.gov/19505943/)]

M2:

[8] M. Pertea, G. M. Pertea, C. M. Antonescu, T.-C. Chang, J. T. Mendell, and S. L. Salzberg, "StringTie enables improved reconstruction of a transcriptome from RNA-seq reads," Nature Biotechnology, vol. 33, no. 3, pp. 290-295, 2015/03/01 2015, doi: 10.1038/nbt.3122. [[paper](https://www.nature.com/articles/nbt.3122)]

[9] A. C. Frazee, G. Pertea, A. E. Jaffe, B. Langmead, S. L. Salzberg, and J. T. Leek, "Flexible isoform-level differential expression analysis with Ballgown," bioRxiv, p. 003665, 2014, doi: 10.1101/003665. [[paper](https://www.biorxiv.org/content/10.1101/003665v1.abstract)]

M3:

[10] Zhen Wei (2020). exomePeak2: Bias Awared Peak Calling and Quantification for MeRIP-Seq. R package version 1.0.0. [[bioc](http://www.bioconductor.org/packages/release/bioc/html/exomePeak2.html)]

[11] S. Heinz, C. Benner, N. Spann, E. Bertolino, Y. C. Lin, P. Laslo et al., "Simple combinations of lineage-determining transcription factors prime cis-regulatory elements required for macrophage and B cell identities," Molecular cell, vol. 38, no. 4, pp. 576-589, 2010. [[paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4763482/)]

[12] Y. Wang, K. Chen, Z. Wei, F. Coenen, J. Su, and J. Meng, "MetaTX: deciphering the distribution of mRNA-related features in the presence of isoform ambiguity, with applications in epitranscriptome analysis," Bioinformatics, 2020, doi: 10.1093/bioinformatics/btaa938. [[paper](https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btaa938/5949013?redirectedFrom=fulltext)]

[13] H. Thorvaldsdóttir, J. T. Robinson, and J. P. Mesirov, "Integrative Genomics Viewer (IGV): high-performance genomics data visualization and exploration," (in eng), Brief Bioinform, vol. 14, no. 2, pp. 178-92, Mar 2013, doi: 10.1093/bib/bbs017. [[paper](https://pubmed.ncbi.nlm.nih.gov/22517427/)]

[14] Q. Liu and R. I. Gregory, "RNAmod: an integrated system for the annotation of mRNA modifications," Nucleic Acids Res, vol. 47, no. W1, pp. W548-W555, 2019, doi: 10.1093/nar/gkz479. [[paper](https://academic.oup.com/nar/article/47/W1/W548/5506862)]

M5:

[15] X. Jiao, B. T. Sherman, D. W. Huang, R. Stephens, M. W. Baseler, H. C. Lane et al., "DAVID-WS: a stateful web service to facilitate gene/protein list analysis," (in eng), Bioinformatics (Oxford, England), vol. 28, no. 13, pp. 1805-1806, 2012, doi: 10.1093/bioinformatics/bts251.[[paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3381967/?report=abstract)]