```shell
# Directory of genome and annotation
$ /data/zhendi/protocol/mm10_ref_genome/Mus_musculus/UCSC/mm10/Sequence/WholeGenomeFasta/genome.fa
$ /data/zhendi/protocol/mm10_ref_genome/Mus_musculus/UCSC/mm10/Sequence/WholeGenomeFasta/genome.fai
$ /data/zhendi/protocol/mm10_ref_genome/Mus_musculus/UCSC/mm10/Annotation/Genes/genes.gtf


# Build genome for alignment
$ hisat2-build -p 20 /data/zhendi/protocol/mm10_ref_genome/Mus_musculus/UCSC/mm10/Sequence/WholeGenomeFasta/genome.fa genome

# write command in hisat_script.sh to run alignment
$ nohup bash hisat_script.sh

# finished exomePeak2

# convert bed to fa
export PATH=$PATH:/home/zhendi/bedtools2/bin

bedtools getfasta -s -fi /data/zhendi/protocol/mm10_ref_genome/Mus_musculus/UCSC/mm10/Sequence/WholeGenomeFasta/genome.fa -bed /data/zhendi/protocol/exomePeak2/data/mm10_re/exomePeak2_output_peakcalling/Mod.bed -split -fo mod.fa &
bedtools getfasta -s -fi /data/zhendi/protocol/mm10_ref_genome/Mus_musculus/UCSC/mm10/Sequence/WholeGenomeFasta/genome.fa -bed /data/zhendi/protocol/exomePeak2/data/mm10_re/exomePeak2_output_diff/DiffMod.bed -split -fo Diffmod.fa &

# streme
export PATH=/home/zhendi/meme-5.3.2/bin:/home/zhendi/meme-5.3.2/libexec/meme-5.3.2:$PATH
streme --dna --objfun de --minw 5 --maxw 10 --o /data/zhendi/protocol/streme_result/mm10/mm_mod --p mod.fa
streme --dna --objfun de --minw 5 --maxw 10 --o /data/zhendi/protocol/streme_result/mm10/mm_diffmod --p Diffmod.fa


# visualize 

```



